test/joystick.js

```diff
@@ -0,0 +1,576 @@
+var MockFirmata = require("./util/mock-firmata"),
+  five = require("../lib/johnny-five.js"),
+  sinon = require("sinon"),
+  Board = five.Board,
+  Joystick = five.Joystick;
+
+var io = new MockFirmata();
+var board = new Board({
+  io: io,
+  debug: false,
+  repl: false
+});
+
+var instance = [{
+  name: "hasAxis",
+}, {
+  name: "x"
+}, {
+  name: "y"
+}];
+
+
+exports["Joystick -- Analog"] = {
+
+  setUp: function(done) {
+
+    this.clock = sinon.useFakeTimers();
+    this.analogRead = sinon.spy(MockFirmata.prototype, "analogRead");
+    this.stick = new Joystick({
+      pins: ["A0", "A1"],
+      board: board
+    });
+
+    done();
+  },
+
+  tearDown: function(done) {
+    this.analogRead.restore();
+    this.clock.restore();
+    done();
+  },
+
+  shape: function(test) {
+    test.expect(instance.length);
+
+    instance.forEach(function(property) {
+      test.notEqual(typeof this.stick[property.name], "undefined");
+    }, this);
+
+    test.done();
+  },
+
+  data: function(test) {
+    var x = this.analogRead.args[0][1];
+    var y = this.analogRead.args[1][1];
+    var spy = sinon.spy();
+
+    test.expect(2);
+
+    this.stick.on("data", spy);
+
+    x(0);
+    y(1023);
+
+    this.clock.tick(200);
+
+    test.ok(spy.calledOnce);
+    test.deepEqual(spy.args[0], [{
+      x: -1,
+      y: 1
+    }]);
+
+    test.done();
+  },
+
+  change: function(test) {
+    test.expect(11);
+
+    var spy = sinon.spy();
+    var x = this.stick.io.analogRead.args[0][1];
+    var y = this.stick.io.analogRead.args[1][1];
+
+    this.stick.on("change", spy);
+
+
+    // FIRST -----------------------------------
+
+
+    x(1023);
+    this.clock.tick(100);
+
+    y(1023);
+    this.clock.tick(100);
+
+    test.equal(Math.round(this.stick.x), 1);
+    test.equal(Math.round(this.stick.y), 1);
+
+    // SECOND -----------------------------------
+
+    x(512);
+    this.clock.tick(100);
+
+    y(512);
+    this.clock.tick(100);
+
+    test.equal(Math.round(this.stick.x), 0);
+    test.equal(Math.round(this.stick.y), 0);
+
+
+    // // THIRD -----------------------------------
+
+    x(0);
+    this.clock.tick(100);
+
+    y(0);
+    this.clock.tick(100);
+
+    test.equal(Math.round(this.stick.x), -1);
+    test.equal(Math.round(this.stick.y), -1);
+
+
+    // // -----------------------------------
+
+    test.equal(spy.callCount, 3);
+
+    test.deepEqual(spy.args[0], [{
+      x: 1,
+      y: 1
+    }]);
+
+    test.equal(Math.round(spy.args[1][0].x), 0);
+    test.equal(Math.round(spy.args[1][0].y), 0);
```

Comment:
I think I need to just revisit the data handling, because it seems wrong to require Math.round